src/rdkafka_telemetry_encode.c

```diff
@@ -183,105 +259,180 @@ void *rd_kafka_telemetry_encode_metrics(rd_kafka_t *rk, size_t *size) {
-        //    attribute.has_value = true;
-        //    attribute.value.which_value =
-        //    opentelemetry_proto_common_v1_AnyValue_string_value_tag;
-        //    attribute.value.value.string_value.funcs.encode = &encode_string;
-        //    attribute.value.value.string_value.arg = "heap";
-        //
-        //    data_point.attributes.funcs.encode = &encode_key_value;
-        //    data_point.attributes.arg = &attribute;
-
-        sum.data_points.funcs.encode = &encode_number_data_point;
-        sum.data_points.arg          = &data_point;
-        sum.aggregation_temporality =
-            opentelemetry_proto_metrics_v1_AggregationTemporality_AGGREGATION_TEMPORALITY_CUMULATIVE;
-        sum.is_monotonic = true;
-
-        metric.which_data = opentelemetry_proto_metrics_v1_Metric_sum_tag;
-        metric.data.sum   = sum;
-
-        metric.description.funcs.encode = &encode_string;
-        metric.description.arg          = metric_description;
-
-        metric.name.funcs.encode = &encode_string;
-        metric.name.arg          = metric_name;
+        metrics = rd_malloc(sizeof(opentelemetry_proto_metrics_v1_Metric *) *
+                            metrics_to_encode_count);
+        data_points =
+            rd_malloc(sizeof(opentelemetry_proto_metrics_v1_NumberDataPoint *) *
+                      metrics_to_encode_count);
+        metric_names = rd_malloc(sizeof(char *) * metrics_to_encode_count);
+        size_t i;
+
+        for (i = 0; i < metrics_to_encode_count; i++) {
+                metrics[i] =
+                    rd_malloc(sizeof(opentelemetry_proto_metrics_v1_Metric));
+                memset(metrics[i], 0,
+                       sizeof(opentelemetry_proto_metrics_v1_Metric));
+                data_points[i] =
+                    (opentelemetry_proto_metrics_v1_NumberDataPoint *)rd_malloc(
+                        sizeof(opentelemetry_proto_metrics_v1_NumberDataPoint));
+                memset(data_points[i], 0,
+                       sizeof(opentelemetry_proto_metrics_v1_NumberDataPoint));
+
+                if (METRICS[metrics_to_encode[i]].is_int) {
+                        data_points[i]->which_value =
+                            opentelemetry_proto_metrics_v1_NumberDataPoint_as_int_tag;
+                        data_points[i]->value.as_int =
+                            METRICS[metrics_to_encode[i]]
+                                .calculate_value(rk)
+                                .intValue;
+
+                } else {
+                        data_points[i]->which_value =
+                            opentelemetry_proto_metrics_v1_NumberDataPoint_as_double_tag;
+                        data_points[i]->value.as_double =
+                            METRICS[metrics_to_encode[i]]
+                                .calculate_value(rk)
+                                .doubleValue;
+                }
+
+                data_points[i]->time_unix_nano = rd_clock();
```

Comment:
We can store a timestamp at the start of this loop so all the metrics encoded at once have the same time stamp, I think that would make sense when we are reading the metrics later (if this is being used somewhere)

Reply:
Makes sense, done.